cupy/functional/piecewise.py

```diff
@@ -0,0 +1,73 @@
+import cupy
+
+from cupy import core
+
+_piecewise_krnl = core.ElementwiseKernel(
+    'U condlist, raw S funclist, int64 xsiz',
+    'raw I prev, raw T y',
+    '''
+        if (condlist && prev[i % xsiz] < i)
+            prev[i % xsiz] = i;
+        __syncthreads();
+        if (prev[i % xsiz] == i)
+            y[i % xsiz] = funclist[i / xsiz];
```

Comment:
The results are undefined when multiple conditions are satisfied. Please avoid such test cases.

Reply:
Could you explain more, please?  You mean the kernel or the test cases. I ensured that when multiple conditions are met, only the latest is considered. So the values are overwritten.

Reply:
Multiple threads will read/write the same memory space `prev[i % xsize]` before syncthreads.

Reply:
I get that multiple threads can't write the same prev[i % xsize], but about reading the same memory memory space ? I mean will replacing syncthreads inside the condition `if (condlist && prev[i % xsiz] < i)` before the writing part resolve this issue ?

Reply:
What about this ?
``` 
        if (condlist){
          __syncthreads();
          if (prev[i % xsiz] < i)
            prev[i % xsiz] = i;
          __syncthreads();
          if (prev[i % xsiz] == i) 
            y[i % xsiz] = funclist[i / xsiz];
        }
```

Reply:
It looks to cause thread divergence.

Reply:
Ok, regarding the original kernel, will atomic operations (reads/ writes or workarounds using cuda) resolve the race condition issue? If not, could you provide suggestions to search/read about ?

Reply:
The atomic operations also cannot resolve the race condition because `__syncthreads` cannot synchronize two threads belonging to different blocks.
I think it's OK to give up parallelizing in the condlist-axis as I commented [here](https://github.com/cupy/cupy/pull/3329#discussion_r423589567).

Reply:
Sorry I overlooked this one :bow: